### PR TITLE
add getPartitionIndex() to the Record<>

### PR DESCRIPTION
**Pull request by dlg99**
### Motivation

Looking at https://github.com/apache/pulsar/pull/9927#discussion_r595714308 there is no way to reliably get partition number.
There is Optional<String> getPartitionId() which returns:
- "{topic}-{partition}" in some Record implementation
- String(partition) in other implementation (parsing back integer is a waste)
- empty Optional (ok) in others

### Modifications

Added `default Optional<Integer> getPartitionIndex()` to the `Record` interface.
Return partition number where appropriate.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: yes - adds new method to `interface Record<T>`
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Review by jerrypeng — APPROVED** [review]

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java [hunk 1]
```diff
@@ -77,6 +77,15 @@
         return Optional.empty();
     }
 
+    /**
+     * Retrieves the partition number if any of the record.
+     *
+     * @return The partition number
+     */
+    default Optional<Integer> getPartitionNumber() {
```
Comment (sijie) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:85]:
```suggestion
    default Optional<Integer> getPartitionId() {
```

Comment (sijie) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:85]: I think parittion id it better than partition number

**Review by sijie — CHANGES_REQUESTED** [review]

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java (hunk 1 above)
Comment (dlg99) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:85]: @sijie I agree but `default Optional<String> getPartitionId() {` already defined on line 76.

File: pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/AbstractKafkaConnectSource.java [hunk 2]
```diff
@@ -183,13 +183,16 @@ public void close() {
...
+        @Getter
+        Optional<Integer> partitionNumber;
 
         KafkaSchemaWrappedSchema keySchema;
 
         KafkaSchemaWrappedSchema valueSchema;
 
         AbstractKafkaSourceRecord(SourceRecord srcRecord) {
             this.destinationTopic = Optional.of("persistent://"+topicNamespace + "/" + srcRecord.topic());
+            this.partitionNumber = Optional.of(srcRecord.kafkaPartition());
```
Comment (eolivelli) [inline pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/AbstractKafkaConnectSource.java:195]: Optional.ofNullable ?

File: pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java [hunk 3]
```diff
@@ -178,6 +178,11 @@ public KafkaRecord(ConsumerRecord<String,?> record, V value, Schema<V> schema) {
             return Optional.of(Integer.toString(record.partition()));
         }
 
+        @Override
+        public Optional<Integer> getPartitionNumber() {
+            return Optional.of(record.partition());
```
Comment (eolivelli) [inline pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java:183]: Optional.ofNullable ?

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarRecord.java [hunk 4]
```diff
@@ -61,6 +61,11 @@
         return Optional.of(topicName);
     }
 
+    @Override
+    public Optional<Integer> getPartitionNumber() {
+        return Optional.of(partition);
```
Comment (eolivelli) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarRecord.java:66]: Optional.ofNullable ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
overall looks good to me
but I left some comments

File: pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java (hunk 3 above)
Comment (dlg99) [reply pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java:183]: `record.partition()` returns int, cannot be null

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarRecord.java (hunk 4 above)
Comment (dlg99) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarRecord.java:66]: `int partition`, cannot be null

**Review by eolivelli — APPROVED** [review]
Lgtm

@sijie you left comments PTAL again

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java (hunk 1 above)
Comment (sijie) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:85]: How about `getPartitionIndex`?

Comment (dlg99) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java]: @sijie renamed

**Review by eolivelli — APPROVED** [review]

**Review by codelipenghui — CHANGES_REQUESTED** [review]
Could you please add some tests for the new method introduced in the PR?

**Review by codelipenghui — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Pull request merged**